### PR TITLE
Fix windows build

### PR DESCRIPTION
**Pull request by Anrock**

### CODE REVIEW

File: src/CabalHelper/Compiletime/Data.hs [hunk 1]
```diff
@@ -13,7 +13,7 @@
 -- You should have received a copy of the GNU Affero General Public License
 -- along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-{-# LANGUAGE TemplateHaskell, ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell, ScopedTypeVariables, CPP #-}
```
Comment (DanielG) [inline src/CabalHelper/Compiletime/Data.hs]: No CPP outside the `Compat` hierarchy please.

File: src/CabalHelper/Compiletime/Compat/Environment.hs [hunk 2]
```diff
@@ -1,6 +1,21 @@
...
+import qualified System.Environment as SE (getEnvironment, setEnv)
+#ifndef mingw32_HOST_OS
+import qualified System.Posix.Env as PE (setEnv)
+#endif
 
 lookupEnv :: String -> IO (Maybe String)
-lookupEnv var = do env <- getEnvironment; return (lookup var env)
+lookupEnv var = do env <- SE.getEnvironment; return (lookup var env)
+
+setEnv :: String -> String -> Bool -> IO ()
```
Comment (DanielG) [inline src/CabalHelper/Compiletime/Compat/Environment.hs]: I don't like the amount of logic in here and the fact that we have to special case dependency bounds on `base`. An easier (and more backwards compatible) way to replace the use of `setEnv` would be to use [`CreateProcess`'s `env` field](https://hackage.haskell.org/package/process-1.6.1.0/docs/System-Process.html#v:env) when spawning cabal commands in the tests.

Comment (DanielG) [reply src/CabalHelper/Compiletime/Compat/Environment.hs]: On second thought that would need a considerable amount of refactoring since we call `cabal` deep in `installCabal`. Ugh how can something as trivial as `setEnv` not be supported prior to `base-4.7`...


In that case just get rid of the `overwrite` argument, it's not like we ever call it with `False` ;)

Also I prefer using fully qualified module names in `Compat` modules, like in https://github.com/DanielG/cabal-helper/blob/master/src/CabalHelper/Compiletime/Compat/Version.hs.

File: cabal-helper.cabal [hunk 3]
```diff
@@ -55,16 +55,21 @@ library
...
                      , mtl           < 2.3  && >= 2.0
                      , process       < 1.7  && >= 1.1.0.1
-                     , unix          < 2.8  && >= 2.5.1.1
                      , semigroupoids < 5.3  && >= 5.2.1
                      , ghc-prim
+  if os(windows)
+    build-depends:     base          < 5    && >= 4.7
+                     , unix-compat
+  else
+    build-depends:     base          < 5    && >= 4.5
```
Comment (DanielG) [inline cabal-helper.cabal]: I'd prefer something like:

```
build-depends:     base < 5 && >= 4.5
  if os(windows)
    build-depends: base        >= 4.7
build-depends:     Cabal < 1.2.3.4 && ...
                   ...
```

Less duplication of information that way and for ease of editing and reading the `if` block should come right after main version bound.

**Review by DanielG — CHANGES_REQUESTED** [review]

**Comment by Anrock** [comment]
@DanielG please review again.

File: cabal-helper.cabal [hunk 4]
```diff
@@ -56,16 +56,22 @@ library
                        Paths_cabal_helper
   ghc-options:         -Wall
   build-depends:       base          < 5    && >= 4.5
-                     , Cabal         < 2.1  && >= 2.0     || < 1.26 && >= 1.14
+  if os(windows)
+    build-depends:     base          < 5    && >= 4.7
```
Comment (DanielG) [inline cabal-helper.cabal]: Please leave off the lower bounds on the `os(windows)` case, they are redundant. All bounds on a package are essentially `&&`ed together so we end up with `< 5 && >= 4.5 && >= 4.7` which resolves to just `< 5 && >= 4.7` but without mentioning the `< 5` twice.

File: cabal-helper.cabal [hunk 5]
```diff
@@ -56,16 +56,22 @@ library
...
+  build-depends:       Cabal         < 2.1  && >= 2.0     || < 1.26 && >= 1.14
                      , directory     < 1.4  && >= 1.1.0.2
                      , filepath      < 1.5  && >= 1.3.0.0
                      , transformers  < 0.6  && >= 0.3.0.0
                      , mtl           < 2.3  && >= 2.0
                      , process       < 1.7  && >= 1.1.0.1
-                     , unix          < 2.8  && >= 2.5.1.1
-                     , semigroupoids < 5.3  && >= 5.2.1
+  if os(windows)
+    build-depends:     unix-compat
```
Comment (DanielG) [inline cabal-helper.cabal]: Add lower and upper bounds to this one though. An underapproximation like `< 0.5 && >= 0.4.3.1` should do to start with.

File: src/CabalHelper/Compiletime/Compat/Environment.hs [hunk 6]
```diff
@@ -1,6 +1,18 @@
...
-import System.Environment
+import qualified System.Environment (getEnvironment, setEnv)
+#ifndef mingw32_HOST_OS
+import qualified System.Posix.Env (setEnv)
+#endif
 
 lookupEnv :: String -> IO (Maybe String)
-lookupEnv var = do env <- getEnvironment; return (lookup var env)
+lookupEnv var = do env <- System.Environment.getEnvironment
+                   return (lookup var env)
```
Comment (DanielG) [inline src/CabalHelper/Compiletime/Compat/Environment.hs]: Unrelated style change, please revert. Don't lump such things in with other changes unless this was causing a build error on windows ;)

**Review by DanielG — CHANGES_REQUESTED** [review]
We're getting there :)

File: src/CabalHelper/Compiletime/Compat/Environment.hs (hunk 6 above)
Comment (Anrock) [reply src/CabalHelper/Compiletime/Compat/Environment.hs]: It's was longer than 80 columns in one line. You want it to be one line or just in different commits?

Comment (DanielG) [reply src/CabalHelper/Compiletime/Compat/Environment.hs]: Oh I read over that `S.E.` qualifier, never mind that then. Do add a line break after the `do` though.

**Comment by Anrock** [comment]
@DanielG can you check again?

**Comment by matsubara0507** [comment]
I can build fixed my cabal-helper in windows (for haskell-ide-engine).
It could be your help: 
https://github.com/DanielG/cabal-helper/compare/master...matsubara0507:fix-unix-lib-for-win-edit

File: cabal-helper.cabal [hunk 7]
```diff
@@ -56,16 +56,22 @@ library
                        Paths_cabal_helper
   ghc-options:         -Wall
   build-depends:       base          < 5    && >= 4.5
-                     , Cabal         < 2.1  && >= 2.0     || < 1.26 && >= 1.14
+  if os(windows)
+    build-depends:     base                 && >= 4.7
```
Comment (DanielG) [inline cabal-helper.cabal]: This doesn't parse at all, did you even build before pushing!?

File: cabal-helper.cabal [hunk 8]
```diff
@@ -56,16 +56,22 @@ library
...
+  build-depends:       Cabal         < 2.1  && >= 2.0     || < 1.26 && >= 1.14
                      , directory     < 1.4  && >= 1.1.0.2
                      , filepath      < 1.5  && >= 1.3.0.0
                      , transformers  < 0.6  && >= 0.3.0.0
                      , mtl           < 2.3  && >= 2.0
                      , process       < 1.7  && >= 1.1.0.1
-                     , unix          < 2.8  && >= 2.5.1.1
-                     , semigroupoids < 5.3  && >= 5.2.1
+  if os(windows)
+    build-depends:     unix-compat   < 0.5  && >= 0.4.3.1
```
Comment (DanielG) [inline cabal-helper.cabal]: Since unix-compat re-exports stuff from unix on non-DOS patforms we might as well only depend on unix-compat and get rid of this conditional entirely. That should also get rid of the need for CPP in the imports.

File: src/CabalHelper/Compiletime/Compat/Environment.hs [hunk 9]
```diff
@@ -1,6 +1,19 @@
...
 
 lookupEnv :: String -> IO (Maybe String)
-lookupEnv var = do env <- getEnvironment; return (lookup var env)
+lookupEnv var =
+  do env <- System.Environment.getEnvironment
+     return (lookup var env)
+
+setEnv :: String -> String -> IO ()
+#ifdef mingw32_HOST_OS
+setEnv = System.Environment.setEnv
```
Comment (DanielG) [inline src/CabalHelper/Compiletime/Compat/Environment.hs:16]: Again this can't (and probably doesn't) compile since you don't import setEnv in the import for `System.Environment` above.

**Review by DanielG — CHANGES_REQUESTED** [review]
Doesn't even compile; please don't waste maintainer time.

File: cabal-helper.cabal (hunk 7 above)
Comment (Anrock) [reply cabal-helper.cabal]: My bad, i'm terribly sorry.

File: cabal-helper.cabal (hunk 8 above)
Comment (Anrock) [reply cabal-helper.cabal]: Unix-compat doesn't export Environment at all, so we can't get rid of unix. I wonder why not use System.Environment.setEnv from base? Is it because it requires base >=4.7?

Comment (DanielG) [reply cabal-helper.cabal]: Yeah, I'd like to support everything back to GHC 7.4, which comes with base-4.5, since there hasn't been a good reason not to so far and doing that covers the most Cabal versions. Having tighter bounds on windos is ok though.

In that case just leave it in.

Comment (Anrock) [reply cabal-helper.cabal]: And i just realized it's probably neccesary to bring unix-compat even on non-win platform, to get rid of ifdefs:
- Either win platform uses 4.7 base for setEnv and unix-compat for everything else and there is ifdefs in Compiletime.Data to use unix-compat on win and unix everywhere else + ifs in .cabal to exclude unix-compat fron non-win dependencies.
- Or we bring unix and unix-compat on both platforms, use unix-compat in Compiletime.data so no ifdefs there and ifs in .cabal. Ifdefs are only left in Compiletime.Compat.

Not sure if i'm right on this.

Comment (DanielG) [reply cabal-helper.cabal]: Using unix-compat on both platforms is fine with me. If it gets rid of conditionals in the cabal file and the source even better.

**Comment by Anrock** [comment]
@DanielG i hope this is last revision. Can't get rid of conditional depend on unix/unix-compat without bumping lower bound of base for non-win too. So i guess ifdefs in Compiletime.Compat and conditional depends have to stay.

**Review by DanielG — APPROVED** [review]

**Comment by DanielG** [comment]
I made a minor adjustment, rebased on master, squashed and pushed this as https://github.com/DanielG/cabal-helper/commit/eb178a5ba4c9432e951b6c206f8baf21257ecb28. Tests should be green: https://gitlab.com/dxld/cabal-helper/pipelines/13656176, they already were in https://gitlab.com/dxld/cabal-helper/pipelines/13653999.

**Comment by DanielG** [comment]
Thanks a bunch!
